modules/csvparser/filterx-func-parse-csv.c

```diff
@@ -103,12 +103,61 @@ _init_scanner(FilterXFunctionParseCSV *self, GList *string_delimiters, GList *co
   if (string_delimiters)
     csv_scanner_options_set_string_delimiters(local_opts, string_delimiters);
 
-  if (cols)
-    csv_scanner_options_set_expected_columns(local_opts, g_list_length(cols));
+  if (num_of_cols)
+    csv_scanner_options_set_expected_columns(local_opts, num_of_cols);
 
   csv_scanner_init(scanner, opts, input);
 }
 
+static inline gboolean
+_maybe_init_columns(FilterXFunctionParseCSV *self, FilterXObject **columns, guint64 *num_of_columns)
+{
+  if (!self->columns)
+    {
+      *columns = NULL;
+      *num_of_columns = 0;
+      return TRUE;
+    }
+
+  *columns = filterx_expr_eval(self->columns);
+  if (!*columns)
+    return FALSE;
+
+  if (!filterx_object_is_type(*columns, &FILTERX_TYPE_NAME(json_array)))
+    {
+      msg_error("list object argument must be a type of json array.",
+                evt_tag_str("current_type", (*columns)->type->name),
+                evt_tag_str("argument_name", FILTERX_FUNC_PARSE_CSV_ARG_NAME_COLUMNS));
+      return FALSE;
+    }
+
+  if (!filterx_object_len(*columns, num_of_columns))
+    return FALSE;
+
+  return TRUE;
+}
+
+static inline gboolean
+_fill_col_object_field(FilterXObject *cols, gint64 index, CSVScanner *scanner, FilterXObject *result)
+{
+  FilterXObject *col = filterx_list_get_subscript(cols, index);
+  const gchar *col_name;
+  gsize col_name_len;
+  filterx_object_extract_string(col, &col_name, &col_name_len);
+
+  FilterXObject *key = filterx_string_new(col_name, col_name_len);
```

Comment:
Why don't we just use `col`? It seems like we have recreated the same string. `set_subscript()` support `message_value`s as well, so there is no need to ensure `FilterXString`.

Reply:
I didn't notice about this, I just refactored this part. 
I've removed the unnecessary string recreation.